crates/ruff_python_parser/src/parser/pattern.rs

```diff
@@ -219,10 +169,24 @@ impl<'src> Parser<'src> {
         let mut rest = None;
 
         self.parse_comma_separated_list(RecoveryContextKind::MatchPatternMapping, |parser| {
+            let mapping_item_start = parser.node_start();
+
             if parser.eat(TokenKind::DoubleStar) {
-                rest = Some(parser.parse_identifier());
+                let identifier = parser.parse_identifier();
+                if rest.is_some() {
+                    parser.add_error(
+                        ParseErrorType::OtherError(
+                            "Only one double star pattern is allowed".to_string(),
+                        ),
+                        parser.node_range(mapping_item_start),
+                    );
+                }
+                // It's not possible to retain multiple double starred patterns because
+                // of the way the mapping node is represented in the grammar. The last
+                // value will always win.
+                rest = Some(identifier);
             } else {
```

Comment:
Hm, this is a problem because we now drop "nodes", meaning the `MatchPatternMapping` range covers nodes that aren't represented in the AST. 

Any chance we could convert the star to a starred expression instead and push a pattern at the right position? If not, then we may need to bail out early here, even if it means that the error recovery isn't as good as it could be (or we have to change the AST representation)

Reply:
> Any chance we could convert the star to a starred expression instead and push a pattern at the right position?

Not really because there's no way to represent a double starred expression in the AST, there's only a `ExprStarred` which is for `*foo`.

> If not, then we may need to bail out early here, even if it means that the error recovery isn't as good as it could be (or we have to change the AST representation)

I could see what bailing would look like but I'd prefer to change the AST representation even it means a bit more work to be done.

Thinking quickly about the AST representation, there are two solutions which comes to my mind:
1. Add a new `Pattern` variant for `**data` and store it in `MatchPatternMapping::patterns` field. This is probably not a good idea
2. Replace the `keys`, `patterns` and `rest` fields with a single `entries` field which is an enum with two kind (`MappingPattern::DoubleStar` and `MappingPattern::KeyValue`)

Reply:
Also, this is happening in a list parsing so might involve using the parser context flags.

Reply:
I'm also leaning on changing the AST (e.g. similar to Pyright's AST). What's important to me is that we don't spend too much time now working on error recovery. I'm also okay marking this as a todo (creating an issue) and following up on this later because the focus now is on replacing the parser, not on perfect error recovery.

Reply:
Yeah, I'd prefer to now to do it now as well.